core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/iterator/ProjectionIterator.java

```diff
@@ -18,32 +22,61 @@
 import org.eclipse.rdf4j.query.algebra.ProjectionElemList;
 import org.eclipse.rdf4j.query.algebra.QueryModelNode;
 import org.eclipse.rdf4j.query.algebra.evaluation.QueryBindingSet;
+import org.eclipse.rdf4j.query.impl.ArrayBindingSet;
 
 public class ProjectionIterator extends ConvertingIteration<BindingSet, BindingSet, QueryEvaluationException> {
 
 	/*-----------*
 	 * Constants *
 	 *-----------*/
-
-	private final Projection projection;
-
-	private final BindingSet parentBindings;
-
-	private final boolean isOuterProjection;
+	private final Function<BindingSet, ? extends BindingSet> converter;
 
 	/*--------------*
 	 * Constructors *
 	 *--------------*/
-
 	public ProjectionIterator(Projection projection, CloseableIteration<BindingSet, QueryEvaluationException> iter,
 			BindingSet parentBindings) throws QueryEvaluationException {
 		super(iter);
-		this.projection = projection;
-		this.parentBindings = parentBindings;
-		this.isOuterProjection = determineOuterProjection();
+		ProjectionElemList pel = projection.getProjectionElemList();
+		if (parentBindings.size() == 0) {
+			if (pel.getElements().size() == 1) {
+				Function<BindingSet, ArrayBindingSet> oneVariableConversion = convertASingleVariable(pel);
```

Comment:
Can we fold the two function implementations into one - perhaps by just counting the number of elements inside the function?

Reply:
We could but then we can't avoid an array length check in the single variable conversion case. On the other hand the single variable case might not go through the projection iterator at all. So I am happy to drop the special one variable case.

Reply:
Third approach was taken. Always use a lambda, and prepare for reuse in the MultiProjectionIterator. I generate three lambda's for the three states possible at construction. No ParentBinding, include all the parentbinding, include the parentbinding value if the source value is null.

Reply:
Looks good to me. The names of the functions are a bit unwieldy, but I must admit I have no inspiration for anything better either, so :+1: